# Extracting Tweets from Tweet IDs

In [ ]:
pip install -U setuptools

Requirement already up-to-date: setuptools in /usr/local/lib/python2.7/dist-packages (44.1.1)


In [ ]:
pip install twython

  Using cached https://files.pythonhosted.org/packages/b7/2c/fb17cb309fccd4e63638715325768bd959b0ad78b23b0cbc0738f5fc8c58/twython-3.8.2.tar.gz
ERROR: Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.


In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import numpy as np 
import pandas as pd 
import twython
from twython import Twython

from tqdm import *

ImportError: ignored

In [ ]:
tweet_ids = pd.read_csv('tweet_ids/2015_Nepal_Earthquake_en/2015_nepal_eq_cf_labels.csv')

In [ ]:
tweet_ids.head()

label              tweet_id
0             other_useful_information  '591902695562170368'
1  infrastructure_and_utilities_damage  '591902695822331904'
2               injured_or_dead_people  '591902695943843840'
3      missing_trapped_or_found_people  '591902696371724288'
4       sympathy_and_emotional_support  '591902696375877632'

I want an additional column, with the actual tweet next to tweet_id

Using Twython

In [ ]:
CONSUMER_KEY = 'I deleted these keys'
CONSUMER_SECRET = 'If you want to generate some'

OAUTH_TOKEN = 'register an app at https://apps.twitter.com/ '
OAUTH_SECRET = 'its super easy!'

In [ ]:
twitter = Twython(CONSUMER_KEY, CONSUMER_SECRET, OAUTH_TOKEN, OAUTH_SECRET)

In [ ]:
test_id = tweet_ids.tweet_id[0][1:-1]

There can be a variety of errors which occur when using this API, such as a suspended user: 

In [ ]:
tweet = twitter.show_status(id=test_id)

TwythonError: Twitter API returned a 403 (Forbidden), User has been suspended.

But I still want the loop to continue if an error is raised, so I'll use Python's 'try, except' to catch any errors

In [ ]:
try: tweet = twitter.show_status(id=test_id)
except: print ("oh rats")

oh rats


In [ ]:
test_id = tweet_ids.tweet_id[1][1:-1] 
try: tweet = twitter.show_status(id=test_id); print tweet['text']
except: print ("oh rats")

RT @DailySabah: #LATEST #Nepal's Kantipur TV shows at least 21 bodies lined up on ground after 7.9 earthquake
http://t.co/opoQLUkYAN http:/…


Excellent. Let's make a new csv file, which has all the tweets filled out and a None if an error was raised. 

In [ ]:
tweet_ids['tweet_texts'] = u''

Note: because of the limited number of calls I could make to Twitter's API, I had to stagger this. 

In [ ]:
for i in tqdm(range(2400, len(tweet_ids))):
    individual_id = tweet_ids.tweet_id.iloc[i][1:-1]
    try: tweet = twitter.show_status(id=individual_id)['text']
    except: tweet = None
    tweet_ids.set_value(i, 'tweet_texts', tweet)

100%|██████████| 618/618 [01:11<00:00,  8.59it/s]


In [ ]:
tweet_ids.to_csv('tweet_ids/2015_Nepal_Earthquake_en/string_filled_tweets.csv', encoding = 'utf-8')

Now, I want to remove all tweets for which no actual tweet could be extracted

In [ ]:
stripped_tweets = tweet_ids[pd.notnull(tweet_ids.tweet_texts)]

In [ ]:
stripped_tweets.to_csv('tweet_ids/2015_Nepal_Earthquake_en/stripped_filled_tweets.csv', encoding = 'utf-8')

I load the data back up to make sure everything is okay: 

In [ ]:
read_tweets = pd.read_csv('tweet_ids/2015_Nepal_Earthquake_en/stripped_filled_tweets.csv', encoding = 'ISO-8859-1')

In [ ]:
len(read_tweets)

2339

So I was able to extract $\frac{2339}{3019} = 77\%$ of the tweets. This should be enough to give this a shot! 